lib/Intlc/Backend/JavaScript/Language.hs

```diff
@@ -0,0 +1,97 @@
+module Intlc.Backend.JavaScript.Language where
+
+import           Intlc.Core (Locale)
+import qualified Intlc.ICU  as ICU
+import           Prelude
+
+type ASTCompiler = Reader Locale
+
+-- | A representation of the output we will be compiling. It's a little verbose
+-- split into these various sum types, but in doing so it's correct by
+-- construction.
+data Stmt = Stmt Text (NonEmpty Expr)
+
+data Expr
+  = TPrint Text
+  | TStr Ref
+  | TNum Ref
+  | TDate Ref ICU.DateTimeFmt
+  | TTime Ref ICU.DateTimeFmt
+  | TApply Ref [Expr]
+  | TMatch Match
+
+data Match = Match Ref MatchCond MatchRet
+
+data MatchCond
+  = LitCond
+  | CardinalPluralRuleCond
+  | OrdinalPluralRuleCond
+
+data MatchRet
+  = LitMatchRet (NonEmpty Branch)
+  | NonLitMatchRet (NonEmpty Branch) Wildcard
+  | RecMatchRet (NonEmpty Branch) Match
+
+newtype Ref = Ref Text
+
+data Branch = Branch Text [Expr]
+newtype Wildcard = Wildcard [Expr]
+
+fromKeyedMsg :: Text -> ICU.Message -> ASTCompiler Stmt
+fromKeyedMsg n (ICU.Static x)   = pure $ Stmt n (pure $ TPrint x)
+fromKeyedMsg n (ICU.Dynamic ys) = Stmt n <$> (fromToken `mapM` ys)
+
+fromToken :: ICU.Token -> ASTCompiler Expr
+fromToken (ICU.Plaintext x)     = pure $ TPrint x
+fromToken (ICU.Interpolation x) = fromArg x
+
+fromArg :: ICU.Arg -> ASTCompiler Expr
+fromArg (ICU.Arg nraw t) =
+  case t of
+    ICU.String             -> pure $ TStr n
+    ICU.Number             -> pure $ TNum n
+    ICU.Date x             -> pure $ TDate n x
+    ICU.Time x             -> pure $ TTime n x
+    ICU.Plural x           -> TMatch <$> fromPlural n x
+    ICU.Select cs (Just w) -> ((TMatch . Match n LitCond) .) . NonLitMatchRet <$> (fromSelectCase `mapM` cs) <*> fromSelectWildcard w
+    ICU.Select cs Nothing  -> TMatch . Match n LitCond . LitMatchRet <$> (fromSelectCase `mapM` cs)
+    ICU.Callback xs        -> TApply n <$> (fromToken `mapM` xs)
+  where n = Ref nraw
+
+fromPlural :: Ref -> ICU.Plural -> ASTCompiler Match
+fromPlural r p = case p of
```

Comment:
Some of this stuff is super verbose and I'm guessing is due to the complexity of what we're trying to solve with plurals. Would be good maybe to go through it over a video so I can understand everything?

Reply:
Certainly, I'm free any time tomorrow if you'd like to video call.

The verbosity here is a good example of what I'd like to improve with #48. The recursion schemes stuff is a bit over my head at the moment, but my first, probably poor, attempt at lenses locally looks like it could reduce boilerplate a lot. Quick example:

```hs
-- before
mapSelectStreams :: (ICU.Stream -> ICU.Stream) -> ICUSelect -> ICUSelect
mapSelectStreams f (xs, mw) = (mapSelectCase f <$> xs, mapSelectWildcard f <$> mw)

mapSelectCase :: (ICU.Stream -> ICU.Stream) -> ICU.SelectCase -> ICU.SelectCase
mapSelectCase f (ICU.SelectCase x ys) = ICU.SelectCase x (f ys)

mapSelectWildcard :: (ICU.Stream -> ICU.Stream) -> ICU.SelectWildcard -> ICU.SelectWildcard
mapSelectWildcard f (ICU.SelectWildcard xs) = ICU.SelectWildcard (f xs)

-- after, first pass
mapSelectStreams :: (ICU.Stream -> ICU.Stream) -> ICUSelect -> ICUSelect
mapSelectStreams f = bimap g g where g = (fmap $ #children %~ f)
```